mon: have mon-specific features

jecluis opened the pull request:
UPDATE: fixed patch set according to @liewegas review. See below for highlights.

This patch set adds the following functionality to the monitors:

1. have a 'features' field in the MonMap, holding the bare minimum features required
by ALL the monitors in the cluster.

2. we now have 'sticky' features, which will be automatically activated - and set on
monmap - once all the quorum members support them. These can not be disabled
or reverted. To be used sparingly.

3. have a 'flags' field in the MonMap, holding (possibly) transient features that are
required by all monitors in the quorum.

4. monitors require a given set of features to join a quorum; these will be assessed
relying on `monmap.features` and `monmap.flags`.

5. any given flag can be set in `monmap.flags` once all the monitors in a quorum
are able to understand said flag. These can also be unset. Setting/unsetting flags
will likely require user intervention (or some action that triggers a flag to be set/unset).

`Signed-off-by: Joao Eduardo Luis <joao@suse.de>`

jecluis commented:
bummer, looks like make check failed on the last commit...

jecluis commented:
yeah, doh. I'll push an updated version of the patch once it goes through my local make check.

jecluis commented:
@liewegas updated patchset.

liewegas commented:
I think this looks great, except for the last flags patch... I'd suggest just leaving that off until we actually have a flag to enable, and at that point add it in.  In particular, I'm not sure it should share the same namespace/bitspace/type as mon_feature_t...

jecluis commented:
@liewegas repushed. I think I addressed all your comments, except the 'flags' patch, which I squashed with the mon/mon_types.h patch, and

1. removed the flags from the feature namespace because there was no point in keeping that around; and
2. renamed 'flags' to 'optional_features' -- alongside with renaming 'features' to 'persistent_features' (no more sticky either, renamed all to persistent).

Also, I updated 'CEPH_FEATURE_JEWEL' to be (1ULL<<57). I have no idea how we're handling the jewel feature bit though, so if this is not desirable I can simply drop the patch. It's mostly meant, in this context, to allow the branch to actually *work*. :)

tchaikov commented on src/mon/mon_types.h:
```diff
@@ -234,4 +236,180 @@ struct C_MonOp : public Context
+  }
+
+  mon_feature_t diff(const mon_feature_t other) const {
+    return mon_feature_t((other.features ^ features) & other.features);
```
@jecluis should be 
```c
(features ^ other.features) & features
```
instead? note, i reordered the transpose "feature" and "other.features" just for readability and consistency.

tchaikov replied:
sorry, scratch that, i thought it's equivalent to `set.difference()` in python. but, imho, it would be easier to other readers also, if the semantic of this function could be changed so that it

> Return a new set with elements in the set that are not in the others.

tchaikov commented on src/mon/mon_types.h:
```diff
@@ -234,4 +236,180 @@ struct C_MonOp : public Context
+  }
+
+  bool contains_all(const mon_feature_t other) const {
+    return !((other.features ^ features) & other.features);
```
instead of reuse the code, why not simply reuse `diff()`? like
```
return !other.diff(*this);
```

tchaikov commented on src/mon/mon_types.h:
```diff
@@ -234,4 +236,180 @@ struct C_MonOp : public Context
+            );
+      }
+
+      constexpr mon_feature_t FLAG_NONE(          (0ULL));
```
can we remove `FLAG_NONE`, i think this is a leftover of `flags` before you change it to `optional_features`?

jecluis replied:
yeah, my bad.

tchaikov commented on src/mon/mon_types.h:
```diff
@@ -234,4 +236,180 @@ struct C_MonOp : public Context
+  return "unknown";
+}
+
+//constexpr mon_feature_t mon_feature_t::FEATURE_JEWEL(10);
```
and can we remove this line also?

jecluis replied:
oops.

tchaikov commented on src/mon/mon_types.h:
```diff
@@ -18,7 +18,9 @@
 #include "common/Formatter.h"
+#include "common/bit_str.h"
 #include "include/Context.h"
+#include "include/stringify.h"
```
`stringify()` or `joinify()` is not used, so we can remove this include?

jecluis replied:
yes.

tchaikov commented on src/mon/MonMap.h:
```diff
@@ -37,6 +38,13 @@ class MonMap {
+  mon_feature_t persistent_features;
+  mon_feature_t optional_features;
+
+  mon_feature_t get_required_features() {
```
could mark this method `const` also.

jecluis replied:
good catch, thanks!

tchaikov commented on src/mon/Monitor.h:
```diff
@@ -227,7 +227,9 @@ class Monitor : public Dispatcher,
   utime_t leader_since;  // when this monitor became the leader, if it is the leader
   utime_t exited_quorum; // time detected as not in quorum; 0 if in
-  uint64_t quorum_features;  ///< intersection of quorum member feature bits
+  uint64_t quorum_con_features;  ///< intersection of quorum member feature bits
```
@jecluis may i know what "con" stands for?

jecluis replied:
from 'connection'.

tchaikov commented on src/mon/MonmapMonitor.cc:
```diff
@@ -104,6 +104,35 @@ void MonmapMonitor::encode_pending(MonitorDBStore::TransactionRef t)
+    return;
+  }
+
+  assert(is_writeable());
```
could put
```c
assert(features.contains_all(pending_map.persistent_features))
``` 
here also.
and use
```c++
mon_feature_t new_features =
  pending_map.persistent_features.diff(features & ceph::features::mon::get_persistent());
```
instead.

tchaikov commented on src/mon/MonMap.h:
```diff
@@ -37,6 +38,13 @@ class MonMap {
   vector<entity_addr_t> rank_addr;
 
+  mon_feature_t persistent_features;
+  mon_feature_t optional_features;
```
is it possible to document `optional_features` in comments here?

tchaikov commented:
@jecluis do we have any test for this feature?

jecluis commented:
@tchaikov none per se; I was hoping issues to be picked up by the upgrade suites. Otherwise, I'm not entirely sure how to test it -- well, maybe except by adding cases to MonMap::generate_test_instances().
